test/integration/framework/master_utils.go

```diff
@@ -0,0 +1,313 @@
+	// Timeout used in benchmarks, to eg: resize an rc
+	DefaultTimeout = 30 * time.Minute
+
+	// Rc manifest used to create pods for benchmarks.
+	// TODO: Convert this to a full path?
+	TestRCManifest = "benchmark-controller.json"
+
+	// Test Namspace, for pods and rcs.
+	TestNS = "test"
+)
+
+// MasterComponents is a control struct for all master components started via NewMasterComponents.
+// TODO: Include all master components (scheduler, nodecontroller).
+// TODO: Reconcile with integration.go, currently the master used there doesn't understand
+// how to restart cleanly, which is required for each iteration of a benchmark. The integration
+// tests also don't make it easy to isolate and turn off components at will.
+type MasterComponents struct {
+	// Raw http server in front of the master
+	ApiServer *httptest.Server
+	// Kubernetes master, contains an embedded etcd helper
+	KubeMaster *master.Master
+	// Restclient used to talk to the kubernetes master
+	RestClient *client.Client
+	// Replication controller manager
+	ControllerManager *controller.ReplicationManager
+	// Channel for stop signals to rc manager
+	rcStopCh chan struct{}
+	// Used to stop master components individually, and via MasterComponents.Stop
+	once sync.Once
+	// Kubernetes etcd helper, has embedded etcd client
+	EtcdHelper *tools.EtcdHelper
+}
+
+// Config is a struct of configuration directives for NewMasterComponents.
+type Config struct {
+	// If nil, a default is used, partially filled configs will not get populated.
+	MasterConfig            *master.Config
+	StartReplicationManager bool
+	// If true, all existing etcd keys are purged before starting master components
+	DeleteEtcdKeys bool
+	// Client throttling qps
+	QPS float32
+	// Client burst qps, also burst replicas allowed in rc manager
+	Burst int
+	// TODO: Add configs for endpoints controller, scheduler etc
+}
+
+// NewMasterComponents creates, initializes and starts master components based on the given config.
+func NewMasterComponents(c *Config) *MasterComponents {
+	m, s, h := startMasterOrDie(c.MasterConfig)
+	// TODO: Allow callers to pipe through a different master url and create a client/start components using it.
+	glog.Infof("Master %+v", s.URL)
+	if c.DeleteEtcdKeys {
+		DeleteAllEtcdKeys()
+	}
+	restClient := client.NewOrDie(&client.Config{Host: s.URL, Version: "v1beta3", QPS: c.QPS, Burst: c.Burst})
+	rcStopCh := make(chan struct{})
+	controllerManager := controller.NewReplicationManager(restClient, c.Burst)
+
+	// TODO: Support events once we can cleanly shutdown an event recorder.
+	controllerManager.SetEventRecorder(&record.FakeRecorder{})
+	if c.StartReplicationManager {
+		go controllerManager.Run(runtime.NumCPU(), rcStopCh)
+	}
+	var once sync.Once
+	return &MasterComponents{
+		ApiServer:         s,
+		KubeMaster:        m,
+		RestClient:        restClient,
+		ControllerManager: controllerManager,
+		rcStopCh:          rcStopCh,
+		EtcdHelper:        h,
+		once:              once,
```

Comment:
Making a final pass here and I don't understand this line-- why is it necessary?